## IBM Data Science - Capstone project

In [1]:
import numpy as np
import pandas as pd

In [2]:
print("Hello capstone project  course!")

Hello capstone project  course!


In [3]:
# Loading the csv
df=pd.read_csv("C://Users//Kargil//Desktop//uk-traffic-accidents-columns.csv")

In [4]:
df.head()

Accident_Index Accident_Severity        Date Day_of_Week  \
0  200501BS00001           Serious  04/01/2005     Tuesday   
1  200501BS00002            Slight  05/01/2005   Wednesday   
2  200501BS00003            Slight  06/01/2005    Thursday   
3  200501BS00004            Slight  07/01/2005      Friday   
4  200501BS00005            Slight  10/01/2005      Monday   

              Light_Conditions  Number_of_Casualties  Number_of_Vehicles  \
0                     Daylight                     1                   1   
1        Darkness - lights lit                     1                   1   
2        Darkness - lights lit                     1                   2   
3                     Daylight                     1                   1   
4  Darkness - lighting unknown                     1                   1   

   Pedestrian_Crossing-Human_Control  Pedestrian_Crossing-Physical_Facilities  \
0                                0.0                                      1.0   
1                                0.0                                      5.0   
2                                0.0                                      0.0   
3                                0.0                                      0.0   
4                                0.0                                      0.0   

  Road_Surface_Conditions  Speed_limit   Time     Weather_Conditions  Year  
0             Wet or damp           30  17:42  Raining no high winds  2005  
1                     Dry           30  17:36     Fine no high winds  2005  
2                     Dry           30   0:15     Fine no high winds  2005  
3                     Dry           30  10:35     Fine no high winds  2005  
4             Wet or damp           30  21:13     Fine no high winds  2005

## Data preprocessing

In [5]:
# Dropping the unwanted columns
df=df.drop(['Pedestrian_Crossing-Human_Control','Pedestrian_Crossing-Physical_Facilities','Accident_Index'],axis=1)

In [6]:
df.columns

Index(['Accident_Severity', 'Date', 'Day_of_Week', 'Light_Conditions',
       'Number_of_Casualties', 'Number_of_Vehicles', 'Road_Surface_Conditions',
       'Speed_limit', 'Time', 'Weather_Conditions', 'Year'],
      dtype='object')

In [7]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1048575 entries, 0 to 1048574
Data columns (total 11 columns):
 #   Column                   Non-Null Count    Dtype 
---  ------                   --------------    ----- 
 0   Accident_Severity        1048575 non-null  object
 1   Date                     1048575 non-null  object
 2   Day_of_Week              1048575 non-null  object
 3   Light_Conditions         1048575 non-null  object
 4   Number_of_Casualties     1048575 non-null  int64 
 5   Number_of_Vehicles       1048575 non-null  int64 
 6   Road_Surface_Conditions  1048575 non-null  object
 7   Speed_limit              1048575 non-null  int64 
 8   Time                     1048475 non-null  object
 9   Weather_Conditions       1048575 non-null  object
 10  Year                     1048575 non-null  int64 
dtypes: int64(4), object(7)
memory usage: 60.0+ MB


In [8]:
# Dropping the null values
df=df.dropna()

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1048475 entries, 0 to 1048574
Data columns (total 11 columns):
 #   Column                   Non-Null Count    Dtype 
---  ------                   --------------    ----- 
 0   Accident_Severity        1048475 non-null  object
 1   Date                     1048475 non-null  object
 2   Day_of_Week              1048475 non-null  object
 3   Light_Conditions         1048475 non-null  object
 4   Number_of_Casualties     1048475 non-null  int64 
 5   Number_of_Vehicles       1048475 non-null  int64 
 6   Road_Surface_Conditions  1048475 non-null  object
 7   Speed_limit              1048475 non-null  int64 
 8   Time                     1048475 non-null  object
 9   Weather_Conditions       1048475 non-null  object
 10  Year                     1048475 non-null  int64 
dtypes: int64(4), object(7)
memory usage: 68.0+ MB


In [10]:
df['Weather_Conditions'].unique()

array(['Raining no high winds', 'Fine no high winds', 'Unknown',
       'Snowing no high winds', 'Other', 'Fine + high winds',
       'Raining + high winds', 'Fog or mist', 'Snowing + high winds',
       'Data missing or out of range'], dtype=object)

In [11]:
df['Road_Surface_Conditions'].unique()

array(['Wet or damp', 'Dry', 'Frost or ice', 'Snow',
       'Flood over 3cm. deep', 'Data missing or out of range'],
      dtype=object)

In [12]:
df['Light_Conditions'].unique()

array(['Daylight', 'Darkness - lights lit', 'Darkness - lighting unknown',
       'Darkness - lights unlit', 'Darkness - no lighting'], dtype=object)

In [13]:
# Representing the days of week as weekday or weekend
df['Day_of_Week'] = df.Day_of_Week.apply(lambda x:'Weekend' if x in ['Saturday','Sunday'] else 'Weekday')

In [14]:
df['Day_of_Week'].unique()

array(['Weekday', 'Weekend'], dtype=object)

In [15]:
df.head()

Accident_Severity        Date Day_of_Week             Light_Conditions  \
0           Serious  04/01/2005     Weekday                     Daylight   
1            Slight  05/01/2005     Weekday        Darkness - lights lit   
2            Slight  06/01/2005     Weekday        Darkness - lights lit   
3            Slight  07/01/2005     Weekday                     Daylight   
4            Slight  10/01/2005     Weekday  Darkness - lighting unknown   

   Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                     1                   1             Wet or damp   
1                     1                   1                     Dry   
2                     1                   2                     Dry   
3                     1                   1                     Dry   
4                     1                   1             Wet or damp   

   Speed_limit   Time     Weather_Conditions  Year  
0           30  17:42  Raining no high winds  2005  
1           30  17:36     Fine no high winds  2005  
2           30   0:15     Fine no high winds  2005  
3           30  10:35     Fine no high winds  2005  
4           30  21:13     Fine no high winds  2005

## Normalizing and encoding the data

In [16]:
# We need to encode the data for applying KNN classifier on it
# Using label encoder for it.
from sklearn.preprocessing import LabelEncoder   
le = LabelEncoder() 
  


In [17]:
# Encoding the day
day=le.fit_transform(df['Day_of_Week'])

In [18]:
# Encoding the light conditions
light=le.fit_transform(df['Light_Conditions'])

In [19]:
# Encoding the road surface conditions
road=le.fit_transform(df['Road_Surface_Conditions'])

In [20]:
# Encoding the weather conditions
weather=le.fit_transform(df['Weather_Conditions'])

In [21]:
# Encoding the accident severity
severity=le.fit_transform(df['Accident_Severity'])

In [22]:
# Zipping all the features needed for classification
features=list(zip(day,light,road,weather))

In [23]:
from sklearn.model_selection import train_test_split


In [24]:
# splitting the dataset for training and testing.
xtrain, xtest, ytrain, ytest = train_test_split(features,severity, test_size=0.6)


## Model deployment

In [37]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
import matplotlib.pyplot as plt


In [26]:
# Using the classifier for k=6 neighbors 
model = KNeighborsClassifier(n_neighbors=6)
model.fit(xtrain,ytrain)


KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=6, p=2,
                     weights='uniform')

In [27]:
# Using the encoded fields for prediction
predicted= model.predict([[0,0,0,0]]) 
print(predicted)

[2]


## Finding the accuracy and optimal K value.

In [30]:
# As the data is big , using only half the data for finding accuracy and the optimal K value.
tt = len(xtrain)
tv = len(xtest)
len(xtrain[int(tt*0.5):]), len(xtest[int(tv*0.5):])


(209695, 314543)

In [34]:
# Finding the best value of K
ks = 10
mean_acc = np.zeros(ks-1)
std_acc = np.zeros(ks-1)

for n in range(4,ks,2):
    neigh = KNeighborsClassifier(n_neighbors = n).fit(xtrain[int(tt*0.5):],ytrain[int(tt*0.5):])
    yhat = neigh.predict(xtest[int(tv*0.5):])
    mean_acc[n-1] = accuracy_score(ytest[int(tv*0.5):],yhat)
    std_acc[n-1] = np.std(yhat==ytest[int(tv*0.5):])/np.sqrt(len(yhat))
print('Best performing K is '+ str(mean_acc.argmax()+1) + ' with an accuracy of ' +str(mean_acc.max()))

Best performing K is 8 with an accuracy of 0.8491462216612673


### The best value of K is 8 and it gives an accuracy of 84.9%